New RPC rewrite causes rpc-using features to hang randomly

### Stacktrace

```
[11:40:22] [Paper Watchdog Thread/ERROR]: ------------------------------
--
4983 | [11:40:22] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH ---
4984 | [11:40:22] [Paper Watchdog Thread/ERROR]: ------------------------------
4985 | [11:40:27] [Paper Watchdog Thread/ERROR]: --- DO NOT REPORT THIS TO PAPER - THIS IS NOT A BUG OR A CRASH  - 1.20.6-148-20f5165 (MC: 1.20.6) ---
4986 | [11:40:27] [Paper Watchdog Thread/ERROR]: The server has not responded for 20 seconds! Creating thread dump
4987 | [11:40:27] [Paper Watchdog Thread/ERROR]: ------------------------------
4988 | [11:40:27] [Paper Watchdog Thread/ERROR]: Server thread dump (Look for plugins here before reporting to Paper!):
4989 | [11:40:27] [Paper Watchdog Thread/ERROR]: ------------------------------
4990 | [11:40:27] [Paper Watchdog Thread/ERROR]: Current Thread: Server thread
4991 | [11:40:27] [Paper Watchdog Thread/ERROR]: 	PID: 79 \| Suspended: false \| Native: false \| State: WAITING
4992 | [11:40:27] [Paper Watchdog Thread/ERROR]: 	Stack:
4993 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/jdk.internal.misc.Unsafe.park(Native Method)
4994 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.util.concurrent.locks.LockSupport.park(LockSupport.java:221)
4995 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1864)
4996 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.util.concurrent.ForkJoinPool.unmanagedBlock(ForkJoinPool.java:4013)
4997 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3961)
4998 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1898)
4999 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2072)
5000 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		app//eu.cloudnetservice.driver.network.rpc.defaults.rpc.DefaultRPC.fireSync(DefaultRPC.java:239)
5001 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		cloudnet-bridge.jar//eu.cloudnetservice.modules.bridge.platform.PlatformPlayerManager$RPC_BRIDGE/0x00007f5638ba6c00.offlinePlayer(Unknown Source)
5002 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		GKCore.jar//dev.robothanzo.gk.core.i18n.I18nPlayer.getName(I18nPlayer.java:46)
5003 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		GKCore.jar//dev.robothanzo.gk.core.i18n.I18nPlayer.update(I18nPlayer.java:131)
5004 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		GKCore.jar//dev.robothanzo.gk.core.Tasks$1.run(Tasks.java:20)
5005 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		org.bukkit.craftbukkit.scheduler.CraftTask.run(CraftTask.java:101)
5006 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		org.bukkit.craftbukkit.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:482)
5007 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1699)
5008 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:467)
5009 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1571)
5010 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1231)
5011 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:323)
5012 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		net.minecraft.server.MinecraftServer$$Lambda/0x00007f5637ee14b0.run(Unknown Source)
5013 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.lang.Thread.runWith(Thread.java:1583)
5014 | [11:40:27] [Paper Watchdog Thread/ERROR]: 		java.base@22.0.1/java.lang.Thread.run(Thread.java:1570)
```


### Actions to reproduce

1. Make a plugin that uses one of the methods that involves RPC communication (in my case: `var offlinePlayer = InjectionLayer.boot().instance(ServiceRegistry.class).firstProvider(PlayerManager.class).offlinePlayer(uuid);`)
2. Run this a few times, the channel may hang unexpectedly and randomly

### CloudNet version

[17.07 00:55:38.860] INFO :
[17.07 00:55:38.861] INFO : CloudNet Blizzard 4.0.0-RC11-SNAPSHOT 202961ff
[17.07 00:55:38.861] INFO : Discord: <https://discord.cloudnetservice.eu/>
[17.07 00:55:38.861] INFO :
[17.07 00:55:38.861] INFO : ClusterId: c6b7bcad-****-401c-****-069f7903704c
[17.07 00:55:38.861] INFO : NodeId: Node-1
[17.07 00:55:38.861] INFO : Head-NodeId: Node-1
[17.07 00:55:38.862] INFO : CPU usage: (P/S) 1.54/12.0/100%
[17.07 00:55:38.862] INFO : Node services memory allocation (U/R/M): 8752/8752/31511 MB
[17.07 00:55:38.862] INFO : Threads: 78
[17.07 00:55:38.862] INFO : Heap usage: 60/256MB
[17.07 00:55:38.862] INFO : JVM: Eclipse Adoptium 22 (OpenJDK 64-Bit Server VM 22.0.1+8)
[17.07 00:55:38.862] INFO : Update Repo: CloudNetService/launchermeta, Update Branch: nightly
[17.07 00:55:38.863] INFO :

### Other

_No response_

### Issue uniqueness

- [X] Yes, this issue is unique. There are no similar issues.

```
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: [17.07 13:58:55.124] ERROR: Exception whilst handling packet BasePacket(channel=1, dataBuf=eu.cloudnetservice.driver.network.netty.buffer.NettyImmutableDataBuf@6b88de16, prioritized=false, creationStamp=2024-07-17T13:58:55.121144363Z, uniqueId=968ff984-312f-4756-adf3-012374c573dc)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: java.lang.IllegalStateException: Exception posting packet from channel 1 to handler eu.cloudnetservice.wrapper.network.listener.PacketServerChannelMessageListener
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.protocol.defaults.DefaultPacketListenerRegistry.handlePacket(DefaultPacketListenerRegistry.java:178)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.protocol.defaults.DefaultPacketListenerRegistry.handlePacket(DefaultPacketListenerRegistry.java:165)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.NettyNetworkHandler.doHandlePacket(NettyNetworkHandler.java:123)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.NettyNetworkHandler.lambda$messageReceived$0(NettyNetworkHandler.java:95)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor$CallerRunsPolicy.rejectedExecution(ThreadPoolExecutor.java:2053)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:841)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at java.base/java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1376)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.NettyNetworkHandler.messageReceived(NettyNetworkHandler.java:95)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.NettyNetworkHandler.messageReceived(NettyNetworkHandler.java:37)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:98)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:455)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.findAndInvokeChannelRead(DefaultChannelHandlerContext.java:445)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:426)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.internal.DelegatingChannelHandlerContext.fireChannelRead(DelegatingChannelHandlerContext.java:113)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder$ByteToMessageDecoderContext.fireChannelRead(ByteToMessageDecoder.java:446)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.codec.NettyPacketDecoder.decode(NettyPacketDecoder.java:77)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:387)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:330)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:204)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:455)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.findAndInvokeChannelRead(DefaultChannelHandlerContext.java:445)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:426)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.internal.DelegatingChannelHandlerContext.fireChannelRead(DelegatingChannelHandlerContext.java:113)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder$ByteToMessageDecoderContext.fireChannelRead(ByteToMessageDecoder.java:446)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.codec.VarInt32FrameDecoder.decode(VarInt32FrameDecoder.java:59)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:387)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:330)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:204)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:455)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.findAndInvokeChannelRead(DefaultChannelHandlerContext.java:445)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:426)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.ChannelHandler.channelRead(ChannelHandler.java:235)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:455)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:838)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.AbstractChannel$ReadSink.processRead(AbstractChannel.java:1975)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.epoll.EpollSocketChannel.epollInReadyBytes(EpollSocketChannel.java:443)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.epoll.EpollSocketChannel.epollInReady(EpollSocketChannel.java:411)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.epoll.AbstractEpollChannel.doReadNow(AbstractEpollChannel.java:305)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.AbstractChannel$ReadSink.readLoop(AbstractChannel.java:2035)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.AbstractChannel.readNow(AbstractChannel.java:910)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.epoll.AbstractEpollChannel.epollInReady(AbstractEpollChannel.java:296)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.epoll.EpollHandler.processReady(EpollHandler.java:452)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.epoll.EpollHandler.run(EpollHandler.java:371)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.SingleThreadEventLoop.runIO(SingleThreadEventLoop.java:192)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.channel.SingleThreadEventLoop.run(SingleThreadEventLoop.java:176)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.concurrent.SingleThreadEventExecutor.lambda$doStartThread$4(SingleThreadEventExecutor.java:774)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.internal.ThreadExecutorMap.lambda$apply$1(ThreadExecutorMap.java:68)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at java.base/java.lang.Thread.run(Thread.java:1570)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: Caused by: eu.cloudnetservice.driver.event.EventListenerException: Error while invoking event listener handle in class dev.robothanzo.gk.room.paper.listeners.ChannelMessageListener
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.event.DefaultRegisteredEventListener.fireEvent(DefaultRegisteredEventListener.java:101)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.event.DefaultEventManager.callEvent(DefaultEventManager.java:87)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.event.EventManager.callEvent(EventManager.java:75)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.wrapper.network.listener.PacketServerChannelMessageListener.handle(PacketServerChannelMessageListener.java:50)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.protocol.defaults.DefaultPacketListenerRegistry.handlePacket(DefaultPacketListenerRegistry.java:176)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	... 48 common frames omitted
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: Caused by: io.netty5.util.concurrent.BlockingOperationException: DefaultPromise@28962248(incomplete)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.concurrent.DefaultPromise.checkDeadLock(DefaultPromise.java:405)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.concurrent.DefaultPromise.checkDeadLock(DefaultPromise.java:400)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.concurrent.DefaultPromise.await(DefaultPromise.java:250)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at io.netty5.util.concurrent.DefaultPromise.sync(DefaultPromise.java:350)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.NettyUtil.awaitFuture(NettyUtil.java:297)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.netty.NettyNetworkChannel.sendPacketSync(NettyNetworkChannel.java:116)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.protocol.defaults.DefaultQueryPacketManager.sendQueryPacket(DefaultQueryPacketManager.java:106)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.DefaultNetworkChannel.sendQueryAsync(DefaultNetworkChannel.java:78)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.rpc.defaults.rpc.DefaultRPC.fire(DefaultRPC.java:285)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.network.rpc.defaults.rpc.DefaultRPC.fireSync(DefaultRPC.java:238)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at GKCore.jar//dev.robothanzo.gk.core.i18n.I18nPlayer.getName(I18nPlayer.java:47)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at GKCore.jar//dev.robothanzo.gk.core.i18n.I18nPlayer.getLegacyDisplayName(I18nPlayer.java:67)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at GKCore.jar//dev.robothanzo.gk.core.i18n.I18nPlayer.getDisplayName(I18nPlayer.java:74)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at GKCore.jar//dev.robothanzo.gk.core.i18n.I18nPlayer.toString(I18nPlayer.java:79)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at GKRoom.jar//dev.robothanzo.gk.room.paper.listeners.ChannelMessageListener.handle(ChannelMessageListener.java:53)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at dev.derklaro.reflexion.internal.handles.MethodHandleAccessorFactory$MethodHandleMethodAccessor.lambda$invoke$1(MethodHandleAccessorFactory.java:396)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at dev.derklaro.reflexion.Result.tryExecute(Result.java:80)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at dev.derklaro.reflexion.internal.handles.MethodHandleAccessorFactory$MethodHandleMethodAccessor.invoke(MethodHandleAccessorFactory.java:394)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at dev.derklaro.reflexion.internal.handles.MethodHandleAccessorFactory$MethodHandleMethodAccessor.invokeWithArgs(MethodHandleAccessorFactory.java:380)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	at eu.cloudnetservice.driver.event.DefaultRegisteredEventListener.fireEvent(DefaultRegisteredEventListener.java:95)
[13:58:55] [multithreadEventLoopGroup-1-1/WARN]: 	... 52 common frames omitted
```
I found this exception appearing from an custom RPC call that my plugin made, maybe this may help?

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/netty/NettyNetworkChannel.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/netty/NettyNetworkChannel.java
@@ -83,21 +83,6 @@ public final class NettyNetworkChannel extends DefaultNetworkChannel implements 
    * {@inheritDoc}
    */
   @Override
-  public void sendPacketSync(@NonNull Packet... packets) {
-    for (var packet : packets) {
-      var future = this.writePacket(packet, false);
-      if (future != null) {
-        NettyUtil.awaitFuture(future);
-      }
-    }
-
-    this.channel.flush();
-  }
-
-  /**
-   * {@inheritDoc}
-   */
-  @Override
   public void sendPacket(@NonNull Packet packet) {
     if (this.channel.executor().inEventLoop()) {
       this.writePacket(packet, true);
@@ -112,8 +97,14 @@ public final class NettyNetworkChannel extends DefaultNetworkChannel implements 
   @Override
   public void sendPacketSync(@NonNull Packet packet) {
     var future = this.writePacket(packet, true);
-    if (future != null) {
-      NettyUtil.awaitFuture(future);
+    if (future != null && !future.executor().inEventLoop()) {
+      // only await the future if we're not currently in the event loop
+      // as this would deadlock the write operations triggered previously
+      try {
+        future.asStage().await();
+      } catch (InterruptedException _) {
+        Thread.currentThread().interrupt(); // reset interrupted state
+      }
     }
   }
 

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/netty/NettyUtil.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/netty/NettyUtil.java
@@ -19,6 +19,8 @@ package eu.cloudnetservice.driver.network.netty;
 import com.google.common.util.concurrent.ThreadFactoryBuilder;
 import eu.cloudnetservice.driver.DriverEnvironment;
 import eu.cloudnetservice.driver.network.netty.buffer.NettyNioBufferReleasingAllocator;
+import eu.cloudnetservice.driver.network.scheduler.NetworkTaskScheduler;
+import eu.cloudnetservice.driver.network.scheduler.ScalingNetworkTaskScheduler;
 import io.netty5.buffer.Buffer;
 import io.netty5.buffer.BufferAllocator;
 import io.netty5.buffer.BufferUtil;
@@ -32,12 +34,7 @@ import io.netty5.handler.codec.DecoderException;
 import io.netty5.handler.ssl.OpenSsl;
 import io.netty5.handler.ssl.SslProvider;
 import io.netty5.util.ResourceLeakDetector;
-import java.util.concurrent.Executor;
 import java.util.concurrent.Executors;
-import java.util.concurrent.LinkedBlockingQueue;
-import java.util.concurrent.RejectedExecutionHandler;
-import java.util.concurrent.ThreadPoolExecutor;
-import java.util.concurrent.TimeUnit;
 import lombok.NonNull;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.Nullable;
@@ -52,12 +49,10 @@ public final class NettyUtil {
 
   private static final int PACKET_DISPATCH_THREADS;
   private static final int NETTY_EVENT_LOOP_THREADS;
-  private static final int PACKET_DISPATCH_QUEUE_SIZE;
 
   private static final SslProvider SELECTED_SSL_PROVIDER;
   private static final NettyTransport SELECTED_NETTY_TRANSPORT;
   private static final BufferAllocator SELECTED_BUFFER_ALLOCATOR;
-  private static final RejectedExecutionHandler DEFAULT_REJECT_HANDLER = new ThreadPoolExecutor.CallerRunsPolicy();
 
   static {
     // check if resource leak detection should be enabled for debugging purposes
@@ -98,7 +93,6 @@ public final class NettyUtil {
     // actual values when the whole context for the allocation is known.
     PACKET_DISPATCH_THREADS = Integer.getInteger("cloudnet.net.packet-dispatch-threads", -1);
     NETTY_EVENT_LOOP_THREADS = Integer.getInteger("cloudnet.net.netty-event-loop-threads", -1);
-    PACKET_DISPATCH_QUEUE_SIZE = Integer.getInteger("cloudnet.net.packet-dispatch-queue-size", -1);
   }
 
   private NettyUtil() {
@@ -124,27 +118,17 @@ public final class NettyUtil {
    * @return a newly created executor for dispatching inbound packets.
    * @throws NullPointerException if the given driver environment is null.
    */
-  public static @NonNull Executor createPacketDispatcher(@NonNull DriverEnvironment driverEnvironment) {
+  public static @NonNull NetworkTaskScheduler createPacketDispatcher(@NonNull DriverEnvironment driverEnvironment) {
     // the maximum thread count that the pool will be allowed to use for packet processing
     // TODO: consider moving the default thread amount for an environment into the environment as a property
     var defaultEnvThreadCount = driverEnvironment.equals(DriverEnvironment.NODE) ? 12 : 4;
     var maximumPoolSize = overriddenCountOrDefault(PACKET_DISPATCH_THREADS, defaultEnvThreadCount);
 
-    // the queue size to use for all packets that cannot be processed due to no processing thread being available
-    // this is the hard limit for the queue, if reached the pool will reject execution of the provided task, which
-    // (in our case due to the setting), means that the netty event loop will take over the processing instead
-    var packetDispatchQueueSize = overriddenCountOrDefault(PACKET_DISPATCH_QUEUE_SIZE, 150);
-    return new ThreadPoolExecutor(
-      maximumPoolSize / 2,
-      maximumPoolSize,
-      30L,
-      TimeUnit.SECONDS,
-      new LinkedBlockingQueue<>(packetDispatchQueueSize),
-      new ThreadFactoryBuilder()
-        .setNameFormat("Packet-Dispatcher-%d")
-        .setThreadFactory(Executors.defaultThreadFactory())
-        .build(),
-      DEFAULT_REJECT_HANDLER);
+    var threadFactory = new ThreadFactoryBuilder()
+      .setNameFormat("Packet-Dispatcher-%d")
+      .setThreadFactory(Executors.defaultThreadFactory())
+      .build();
+    return new ScalingNetworkTaskScheduler(threadFactory, maximumPoolSize);
   }
 
   /**

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/netty/NettyUtil.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/netty/NettyUtil.java
@@ -32,9 +32,6 @@ import io.netty5.handler.codec.DecoderException;
 import io.netty5.handler.ssl.OpenSsl;
 import io.netty5.handler.ssl.SslProvider;
 import io.netty5.util.ResourceLeakDetector;
-import io.netty5.util.concurrent.Future;
-import java.util.concurrent.CancellationException;
-import java.util.concurrent.CompletionException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.Executors;
 import java.util.concurrent.LinkedBlockingQueue;
@@ -275,29 +272,6 @@ public final class NettyUtil {
       return 3;
     } else {
       return 4;
-    }
-  }
-
-  /**
-   * Waits for the given future to complete, either returning the same future instance as given (but completed) or
-   * rethrowing all exceptions that occurred during completion. This method throws an IllegalThreadStateException if the
-   * current thread was interrupted during the future computation.
-   *
-   * @param future the future to wait for.
-   * @param <T>    the type of data returned by the future.
-   * @return the same future as given to the method, but completed.
-   * @throws NullPointerException        if the given future is null.
-   * @throws CancellationException       if the computation was cancelled
-   * @throws CompletionException         if the computation threw an exception.
-   * @throws IllegalThreadStateException if the current thread was interrupted during computation.
-   */
-  public static @NonNull <T> Future<T> awaitFuture(@NonNull Future<T> future) {
-    try {
-      // await the future and rethrow exceptions if any occur
-      return future.asStage().sync().future();
-    } catch (InterruptedException exception) {
-      Thread.currentThread().interrupt(); // reset the interrupted state of the thread
-      throw new IllegalThreadStateException();
     }
   }
 

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/netty/client/NettyNetworkClient.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/netty/client/NettyNetworkClient.java
@@ -27,6 +27,7 @@ import eu.cloudnetservice.driver.network.netty.NettyUtil;
 import eu.cloudnetservice.driver.network.protocol.Packet;
 import eu.cloudnetservice.driver.network.protocol.PacketListenerRegistry;
 import eu.cloudnetservice.driver.network.protocol.defaults.DefaultPacketListenerRegistry;
+import eu.cloudnetservice.driver.network.scheduler.NetworkTaskScheduler;
 import eu.cloudnetservice.driver.network.ssl.SSLConfiguration;
 import io.netty5.bootstrap.Bootstrap;
 import io.netty5.channel.ChannelOption;
@@ -67,7 +68,7 @@ public class NettyNetworkClient implements NetworkClient {
   protected final PacketListenerRegistry packetRegistry = new DefaultPacketListenerRegistry();
 
   protected final EventManager eventManager;
-  protected final Executor packetDispatcher;
+  protected final NetworkTaskScheduler packetDispatcher;
   protected final Callable<NetworkChannelHandler> handlerFactory;
 
   /**
@@ -202,6 +203,7 @@ public class NettyNetworkClient implements NetworkClient {
   @Override
   public void close() {
     this.closeChannels();
+    this.packetDispatcher.shutdown();
     this.eventLoopGroup.shutdownGracefully();
   }
 

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/netty/server/NettyNetworkServer.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/netty/server/NettyNetworkServer.java
@@ -28,6 +28,7 @@ import eu.cloudnetservice.driver.network.netty.NettyUtil;
 import eu.cloudnetservice.driver.network.protocol.Packet;
 import eu.cloudnetservice.driver.network.protocol.PacketListenerRegistry;
 import eu.cloudnetservice.driver.network.protocol.defaults.DefaultPacketListenerRegistry;
+import eu.cloudnetservice.driver.network.scheduler.NetworkTaskScheduler;
 import eu.cloudnetservice.driver.network.ssl.SSLConfiguration;
 import io.netty5.bootstrap.ServerBootstrap;
 import io.netty5.channel.ChannelOption;
@@ -75,7 +76,7 @@ public class NettyNetworkServer implements NetworkServer {
   protected final PacketListenerRegistry packetRegistry = new DefaultPacketListenerRegistry();
 
   protected final EventManager eventManager;
-  protected final Executor packetDispatcher;
+  protected final NetworkTaskScheduler packetDispatcher;
   protected final Callable<NetworkChannelHandler> handlerFactory;
 
   /**
@@ -236,11 +237,11 @@ public class NettyNetworkServer implements NetworkServer {
   @Override
   public void close() {
     this.closeChannels();
-
     for (var entry : this.channelFutures.values()) {
       entry.cancel();
     }
 
+    this.packetDispatcher.shutdown();
     this.bossEventLoopGroup.shutdownGracefully();
     this.workerEventLoopGroup.shutdownGracefully();
   }

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/netty/server/NettyNetworkServer.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/netty/server/NettyNetworkServer.java
@@ -293,16 +293,6 @@ public class NettyNetworkServer implements NetworkServer {
    * {@inheritDoc}
    */
   @Override
-  public void sendPacketSync(@NonNull Packet... packets) {
-    for (var channel : this.channels) {
-      channel.sendPacketSync(packets);
-    }
-  }
-
-  /**
-   * {@inheritDoc}
-   */
-  @Override
   public void closeChannels() {
     var iterator = this.channels.iterator();
     while (iterator.hasNext()) {

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/protocol/PacketSender.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/protocol/PacketSender.java
@@ -53,17 +53,4 @@ public interface PacketSender {
       this.sendPacket(packet);
     }
   }
-
-  /**
-   * Sends all the given packets to the associated target, waiting for the packet writes to the channel to complete
-   * before resuming the current thread.
-   *
-   * @param packets the packets to send.
-   * @throws NullPointerException if the packets are null.
-   */
-  default void sendPacketSync(@NonNull Packet... packets) {
-    for (var packet : packets) {
-      this.sendPacketSync(packet);
-    }
-  }
 }

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/FallbackRejectionHandler.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/FallbackRejectionHandler.java
@@ -1,0 +1,49 @@
+/*
+ * Copyright 2019-2024 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.scheduler;
+
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.RejectedExecutionHandler;
+import java.util.concurrent.ThreadPoolExecutor;
+import lombok.NonNull;
+
+/**
+ * A rejected execution handler that starts the rejected task in the provided fallback executor.
+ *
+ * @param fallbackExecutor the fallback executor to start rejected starts in.
+ * @since 4.0
+ */
+record FallbackRejectionHandler(@NonNull ThreadPoolExecutor fallbackExecutor) implements RejectedExecutionHandler {
+
+  /**
+   * Schedules the given rejected task in the fallback executor.
+   *
+   * @param task              the task requested to be executed.
+   * @param rejectingExecutor the executor that failed to schedule the task.
+   * @throws NullPointerException if the given task or rejecting executor is null.
+   */
+  @Override
+  public void rejectedExecution(@NonNull Runnable task, @NonNull ThreadPoolExecutor rejectingExecutor) {
+    if (!rejectingExecutor.isShutdown() && !this.fallbackExecutor.isShutdown()) {
+      try {
+        this.fallbackExecutor.execute(task);
+      } catch (RejectedExecutionException _) {
+        // ignore this, must be due to a concurrent shutdown call
+      }
+    }
+  }
+}

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/NetworkTaskScheduler.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/NetworkTaskScheduler.java
@@ -1,0 +1,34 @@
+/*
+ * Copyright 2019-2024 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.scheduler;
+
+import java.util.concurrent.Executor;
+
+/**
+ * A scheduler for tasks that are triggered by incoming network calls and should be handled non-blocking (for execute
+ * invocations) with the lowest latency possible.
+ *
+ * @since 4.0
+ */
+public interface NetworkTaskScheduler extends Executor {
+
+  /**
+   * Triggers a shutdown operation on this scheduler, interrupting all currently running tasks and preventing new tasks
+   * from being scheduled.
+   */
+  void shutdown();
+}

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/ScalingNetworkTaskScheduler.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/ScalingNetworkTaskScheduler.java
@@ -37,9 +37,9 @@ import lombok.NonNull;
  */
 public final class ScalingNetworkTaskScheduler implements NetworkTaskScheduler {
 
-  private static final int BASE_QUEUE_TIMEOUT = 500;
-  private static final int MINIMUM_QUEUE_TIMEOUT = 100;
-  private static final double QUEUE_TIMOUT_DECAY_FACTOR = 0.99;
+  private static final int BASE_QUEUE_TIMEOUT = 150;
+  private static final int MINIMUM_QUEUE_TIMEOUT = 50;
+  private static final double QUEUE_TIMOUT_DECAY_FACTOR = 0.95;
 
   private final AtomicBoolean active;
   private final ThreadPoolExecutor coreExecutor;

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/ScalingNetworkTaskScheduler.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/ScalingNetworkTaskScheduler.java
@@ -1,0 +1,133 @@
+/*
+ * Copyright 2019-2024 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.scheduler;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.LongSupplier;
+import lombok.NonNull;
+
+/**
+ * An implementation of a network task scheduler that slowly starts more threads depending on the count of tasks that
+ * couldn't be scheduled into the core executor service. If more and more tasks are starting to come in and the core
+ * thread pool is not able to keep up, the pace of starting new threads increases. Initially the time between thread
+ * starts is 2.5 seconds, and it can go all the way down to 100ms if needed (starting at around 160 tasks waiting to be
+ * scheduled).
+ *
+ * @since 4.0
+ */
+public final class ScalingNetworkTaskScheduler implements NetworkTaskScheduler {
+
+  private static final int BASE_QUEUE_TIMEOUT = 500;
+  private static final int MINIMUM_QUEUE_TIMEOUT = 100;
+  private static final double QUEUE_TIMOUT_DECAY_FACTOR = 0.99;
+
+  private final AtomicBoolean active;
+  private final ThreadPoolExecutor coreExecutor;
+  private final ThreadPoolExecutor fallbackExecutor;
+  private final TaskSchedulingAction taskScheduler;
+
+  /**
+   * Constructs a new scaling task executor.
+   *
+   * @param threadFactory       the thread factory to use when the underlying executors create new threads.
+   * @param maximumCorePoolSize the maximum threads that the core executor can use.
+   * @throws NullPointerException     if the given thread factory is null.
+   * @throws IllegalArgumentException if the given maximum pool size is smaller than 0.
+   */
+  public ScalingNetworkTaskScheduler(@NonNull ThreadFactory threadFactory, int maximumCorePoolSize) {
+    this.active = new AtomicBoolean(true);
+
+    // use a thread pool with an infinite possible amount of threads as a fallback, but only
+    // keep these threads alive for a few seconds. this only the last-resort executor and shouldn't
+    // have a lot of resources allocated for no reason. note that there is no rejection handler
+    // set, this executor can technically not reject tasks because it can spawn an infinite
+    // amount of threads
+    this.fallbackExecutor = new ThreadPoolExecutor(
+      0,
+      Integer.MAX_VALUE,
+      5L,
+      TimeUnit.SECONDS,
+      new SynchronousQueue<>(),
+      threadFactory);
+
+    // construct the calculator for timeouts when adding tasks into the core executor,
+    // which depends on the count of unscheduled (more unscheduled tasks = smaller timeout)
+    LinkedBlockingQueue<Runnable> unscheduledTaskQueue = new LinkedBlockingQueue<>();
+    LongSupplier timeoutSupplier = () -> {
+      var scheduledTasks = unscheduledTaskQueue.size();
+      var timeout = BASE_QUEUE_TIMEOUT * Math.pow(QUEUE_TIMOUT_DECAY_FACTOR, scheduledTasks);
+      return (long) Math.clamp(timeout, MINIMUM_QUEUE_TIMEOUT, BASE_QUEUE_TIMEOUT);
+    };
+
+    // construct the core executor for incoming packets, which should be used primarily
+    // for scheduling actions. if this scheduler runs out of available threads (for example
+    // if they are all blocked) after a short timeout the tasks are scheduled into a fallback
+    // scheduler which has no size limits and takes over the work to not possibly run into
+    // an issue with too many blocked threads
+    TimedSynchronousQueue<Runnable> workQueue = new TimedSynchronousQueue<>(timeoutSupplier);
+    var fallbackDelegatingRejectionHandler = new FallbackRejectionHandler(this.fallbackExecutor);
+    this.coreExecutor = new ThreadPoolExecutor(
+      maximumCorePoolSize / 2,
+      maximumCorePoolSize,
+      30L,
+      TimeUnit.SECONDS,
+      workQueue,
+      threadFactory,
+      fallbackDelegatingRejectionHandler);
+    workQueue.parentExecutor(this.coreExecutor); // allows the queue to make some optimizations
+
+    // construct and start the actual task scheduling action, uses a virtual
+    // thread here as the action blocks most of the time waiting for tasks
+    // to be executed
+    this.taskScheduler = new TaskSchedulingAction(this.coreExecutor, unscheduledTaskQueue);
+    Thread.ofVirtual().name("NetworkTaskScheduler").start(this.taskScheduler);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void shutdown() {
+    if (this.active.compareAndSet(true, false)) {
+      this.coreExecutor.shutdownNow();
+      this.fallbackExecutor.shutdownNow();
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void execute(@NonNull Runnable command) {
+    if (!this.active.get()) {
+      // the underlying scheduler were also shut down, there is no point
+      // in even trying to schedule a new task in them
+      throw new RejectedExecutionException("scheduler was shut down");
+    }
+
+    // scheduling directly into the core executor would be a blocking operation
+    // so we use this "man-in-the-middle" action that does the scheduling into
+    // the core executor without blocking the caller of this method
+    this.taskScheduler.scheduleTask(command);
+  }
+}

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/TaskSchedulingAction.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/TaskSchedulingAction.java
@@ -1,0 +1,90 @@
+/*
+ * Copyright 2019-2024 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.scheduler;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.ThreadPoolExecutor;
+import lombok.NonNull;
+
+/**
+ * The task scheduler runnable that is responsible for scheduling incoming tasks into the target executor service. Due
+ * to the fact that the scheduling action could be blocking, this action should be scheduled in a separate thread.
+ *
+ * @since 4.0
+ */
+final class TaskSchedulingAction implements Runnable {
+
+  private final ThreadPoolExecutor targetExecutor;
+  private final LinkedBlockingQueue<Runnable> unscheduledTasks;
+
+  /**
+   * Constructs a new task scheduling action.
+   *
+   * @param targetExecutor       the executor to execute unscheduled tasks.
+   * @param unscheduledTaskQueue the queue for unscheduled tasks.
+   * @throws NullPointerException if the given target executor or unscheduled task queue is null.
+   */
+  public TaskSchedulingAction(
+    @NonNull ThreadPoolExecutor targetExecutor,
+    @NonNull LinkedBlockingQueue<Runnable> unscheduledTaskQueue
+  ) {
+    this.targetExecutor = targetExecutor;
+    this.unscheduledTasks = unscheduledTaskQueue;
+  }
+
+  /**
+   * Schedules tasks that were added to the unscheduled tasks queue into the target executor. This operation blocks
+   * until a task is available in the tasks queue.
+   */
+  @Override
+  public void run() {
+    while (true) {
+      try {
+        var nextTask = this.unscheduledTasks.take();
+        if (this.targetExecutor.isShutdown()) {
+          // the shutdown call on the target executor might happen
+          // while we're waiting for an unscheduled task to become
+          // available, therefore we need to check after to prevent
+          // scheduling in an executor that was shut down already
+          break;
+        }
+
+        // this call might block for a bit, depending on the queue used by the
+        // target executor (therefore we cannot call it directly on the caller)
+        this.targetExecutor.execute(nextTask);
+      } catch (InterruptedException _) {
+        Thread.currentThread().interrupt(); // reset interrupted state
+        break;
+      } catch (RejectedExecutionException _) {
+        // the executor rejected the execution, this should
+        // usually be handled otherwise... but just to be sure
+        // that this occurrence wouldn't blow up this scheduler
+      }
+    }
+  }
+
+  /**
+   * Schedules the given task for execution.
+   *
+   * @param task the task to execute.
+   * @throws NullPointerException if the given task is null.
+   */
+  public void scheduleTask(@NonNull Runnable task) {
+    this.unscheduledTasks.offer(task);
+  }
+}

--- a/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/TimedSynchronousQueue.java
+++ b/driver/src/main/java/eu/cloudnetservice/driver/network/scheduler/TimedSynchronousQueue.java
@@ -1,0 +1,125 @@
+/*
+ * Copyright 2019-2024 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.scheduler;
+
+import com.google.common.base.Preconditions;
+import java.util.concurrent.SynchronousQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.LongSupplier;
+import lombok.NonNull;
+
+/**
+ * An implementation of a synchronous queue that always offers tasks to itself with a timeout instead of returning
+ * instantly if that is not possible. The queue also makes some more assumptions when the parent executor service is
+ * provided to achieve the maximum scaling of the parent thread pool.
+ *
+ * @param <E> the type of elements held in this queue
+ * @since 4.0
+ */
+final class TimedSynchronousQueue<E> extends SynchronousQueue<E> {
+
+  private final LongSupplier timeoutSupplier;
+  private ThreadPoolExecutor parentExecutor;
+
+  /**
+   * Constructs a new instance of this queue using the given timeout supplier.
+   *
+   * @param timeoutSupplier the supplier for timeouts used when actually needing to queue elements.
+   * @throws NullPointerException if the given timeout supplier is null.
+   */
+  public TimedSynchronousQueue(@NonNull LongSupplier timeoutSupplier) {
+    super(true); // let waiting threads contend in FIFO order
+    this.timeoutSupplier = timeoutSupplier;
+  }
+
+  /**
+   * Sets the parent thread pool executor that is associated with the queue unless another thread pool executor is
+   * already associated with it.
+   *
+   * @param parentExecutor the parent executor to use for this queue.
+   * @throws NullPointerException  if the given parent executor is null.
+   * @throws IllegalStateException if a parent executor was already set.
+   */
+  public void parentExecutor(@NonNull ThreadPoolExecutor parentExecutor) {
+    Preconditions.checkState(this.parentExecutor == null, "parent executor already set");
+    this.parentExecutor = parentExecutor;
+  }
+
+  /**
+   * Tries to offer the given element into this queue with a timeout provided by the owner of this queue. If the queue
+   * has information about the associated thread pool this method makes some additional considerations to ensure scaling
+   * the thread pool to its maximum capacity first.
+   *
+   * @param element the element to add into this queue.
+   * @return true if the element was added into this queue, false otherwise.
+   * @throws NullPointerException if the given element is null.
+   */
+  @Override
+  public boolean offer(@NonNull E element) {
+    var executor = this.parentExecutor;
+    if (executor != null) {
+      // reject immediately if the parent executor is already shut down, no need to try
+      // to queue it as there won't be any thread to pick up the given task anyway
+      if (executor.isShutdown()) {
+        return false;
+      }
+
+      // check if there are idling threads in the thread pool at the moment. if that is the case
+      // try to enqueue the element for 50ms, which should be more than enough for one of the idling
+      // workers to wake up and poll the task from this queue.
+      // note that the returned active count is an approximation, therefore it could be that there are
+      // no free threads anymore when the actual offer call happens, hence the timeout
+      if (executor.getActiveCount() < executor.getPoolSize()) {
+        var polledByExecutor = this.offerSafe(element, 50);
+        if (polledByExecutor) {
+          return true;
+        }
+      }
+
+      // check if the threads of the executor are already maxed out. TPE prefers to queue
+      // tasks instead of spawning a new thread for them, this basically forces the executor
+      // to go ahead and spawn a new one as we aren't at the maximum pool size yet
+      if (executor.getPoolSize() < executor.getMaximumPoolSize()) {
+        return false;
+      }
+    }
+
+    // either no information is available about the thread pool or the thread pool is completely
+    // busy at the moment. try to queue the task with a timeout, if this doesn't work the task
+    // will be rejected which we can handle somewhere else
+    var offerTimeout = this.timeoutSupplier.getAsLong();
+    return this.offerSafe(element, offerTimeout);
+  }
+
+  /**
+   * Tries to offer the given element into this queue with the given timeout millis.
+   *
+   * @param element       the element to offer into this queue.
+   * @param timeoutMillis the timeout in milliseconds of the offer operation.
+   * @return true if some other thread polled the element within the timeout millis, false otherwise.
+   * @throws NullPointerException if the given element is null.
+   */
+  private boolean offerSafe(@NonNull E element, long timeoutMillis) {
+    try {
+      return super.offer(element, timeoutMillis, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException _) {
+      Thread.currentThread().interrupt(); // reset interrupted state
+      return false;
+    }
+  }
+}

--- a/driver/src/test/java/eu/cloudnetservice/driver/network/netty/NettyUtilTest.java
+++ b/driver/src/test/java/eu/cloudnetservice/driver/network/netty/NettyUtilTest.java
@@ -21,44 +21,10 @@ import io.netty5.buffer.BufferAllocator;
 import io.netty5.channel.MultithreadEventLoopGroup;
 import io.netty5.handler.ssl.OpenSsl;
 import io.netty5.handler.ssl.SslProvider;
-import java.util.concurrent.LinkedBlockingQueue;
-import java.util.concurrent.ThreadPoolExecutor;
-import java.util.concurrent.TimeUnit;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 
 public class NettyUtilTest {
-
-  @Test
-  void testPacketDispatcherThreadCount() {
-    {
-      // test node dispatcher
-      var packetDispatcher = NettyUtil.createPacketDispatcher(DriverEnvironment.NODE);
-      var tpe = Assertions.assertInstanceOf(ThreadPoolExecutor.class, packetDispatcher);
-      Assertions.assertEquals(12, tpe.getMaximumPoolSize());
-      Assertions.assertEquals(6, tpe.getCorePoolSize());
-      Assertions.assertEquals(30, tpe.getKeepAliveTime(TimeUnit.SECONDS));
-      Assertions.assertDoesNotThrow(() -> tpe.getRejectedExecutionHandler().rejectedExecution(() -> {
-      }, tpe));
-
-      var dispatchQueue = Assertions.assertInstanceOf(LinkedBlockingQueue.class, tpe.getQueue());
-      Assertions.assertEquals(150, dispatchQueue.remainingCapacity());
-    }
-
-    {
-      // test wrapper dispatcher
-      var packetDispatcher = NettyUtil.createPacketDispatcher(DriverEnvironment.WRAPPER);
-      var tpe = Assertions.assertInstanceOf(ThreadPoolExecutor.class, packetDispatcher);
-      Assertions.assertEquals(4, tpe.getMaximumPoolSize());
-      Assertions.assertEquals(2, tpe.getCorePoolSize());
-      Assertions.assertEquals(30, tpe.getKeepAliveTime(TimeUnit.SECONDS));
-      Assertions.assertDoesNotThrow(() -> tpe.getRejectedExecutionHandler().rejectedExecution(() -> {
-      }, tpe));
-
-      var dispatchQueue = Assertions.assertInstanceOf(LinkedBlockingQueue.class, tpe.getQueue());
-      Assertions.assertEquals(150, dispatchQueue.remainingCapacity());
-    }
-  }
 
   @Test
   void testBossEventLoopGroupCreation() {

--- a/driver/src/test/java/eu/cloudnetservice/driver/network/scheduler/ScalingNetworkTaskSchedulerTest.java
+++ b/driver/src/test/java/eu/cloudnetservice/driver/network/scheduler/ScalingNetworkTaskSchedulerTest.java
@@ -1,0 +1,51 @@
+/*
+ * Copyright 2019-2024 CloudNetService team & contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.cloudnetservice.driver.network.scheduler;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+public class ScalingNetworkTaskSchedulerTest {
+
+  @Test
+  void testScalingOfSchedulerIfCoreThreadsAreBlocked() throws InterruptedException {
+    var executedTasks = new CountDownLatch(5);
+    var scheduler = new ScalingNetworkTaskScheduler(Executors.defaultThreadFactory(), 2);
+
+    // block the two core threads of the scheduler for an infinite amount of time
+    Runnable blockingAction = () -> {
+      try {
+        Thread.sleep(Long.MAX_VALUE);
+      } catch (InterruptedException _) {
+      }
+    };
+    scheduler.execute(blockingAction);
+    scheduler.execute(blockingAction);
+
+    // schedule more tasks into the thread pool than the wrapped core thread pool
+    // can handle, this triggers the tasks to be put into the fallback scheduler
+    for (var taskId = 0; taskId < 5; taskId++) {
+      scheduler.execute(executedTasks::countDown);
+    }
+
+    var countReachedZero = executedTasks.await(30, TimeUnit.SECONDS);
+    Assertions.assertTrue(countReachedZero);
+  }
+}

